server/src/main/scala/org/broadinstitute/dsp/workbench/welder/server/ObjectService.scala

```diff
@@ -30,21 +44,104 @@ class ObjectService(googleStorageService: GoogleStorageService[IO]) extends Http
+                            syncStatus = if (fileBody.length == 0) SyncStatus.RemoteNotFound
+                              else if (calculatedCrc32 == crc32) SyncStatus.Live
+                              else SyncStatus.Desynchronized
+                            lastEditedBy = userDefinedMetadata.get("lastEditedBy").map(WorkbenchEmail)
+                            lastEditedTime <- userDefinedMetadata.get("lastEditedTime").flatTraverse[IO, Instant] {
+                              str =>
+                                Either.catchNonFatal(str.toLong) match {
+                                  case Left(t) => logger.warn(s"Failed to convert $str to epoch millis") *> IO.pure(None)
+                                  case Right(s) => IO.pure(Some(Instant.ofEpochMilli(s)))
+                                }
+                            }
+                          } yield MetadataResponse.Found(isLinked, syncStatus, lastEditedBy, lastEditedTime, sl.cloudStorageDirectory, true, sl) //TODO: fix isExecutionMode
+                      }
+          } yield result
+      }
+    } yield res
 
-  def safeDelocalize(req: SafeDelocalize): IO[Unit] = ???
+  def safeDelocalize(req: SafeDelocalize): IO[Unit] = {
+    for {
+      traceId <- IO(TraceId(randomUUID()))
+      storageLinks <- readJsonFileToA[IO, List[StorageLink]](pathToStorageLinks).compile.lastOrError
+      storageLink = storageLinks.find(_.localBaseDirectory == req.localObjectPath)
+      _ <- storageLink.fold[IO[Unit]](IO.raiseError(StorageLinkNotFoundException(s"No storage link found for ${req.localObjectPath}"))) {
+        sl =>
+          for {
+            metadata <- checkMetadata(GetMetadataRequest(req.localObjectPath))
+            _ <- metadata match {
+              case x: MetadataResponse.RemoteNotFound => delocalize(req.localObjectPath, sl.cloudStorageDirectory, traceId)
+              case meta: MetadataResponse.Found =>
+                for {
+                  _ <- if(meta.syncStatus == SyncStatus.Live) {
+                    delocalize(req.localObjectPath, sl.cloudStorageDirectory, traceId)
+                  } else {
+                    val msg = s"File is outdated(${meta.syncStatus}), hence not delocalizing"
+                    logger.info(msg) *> IO.raiseError(FileOutOfSyncException(msg))
+                  }
+                } yield ()
+            }
+          } yield ()
+      }
+    } yield ()
+  }
+
+  private def delocalize(localObjectPath: LocalObjectPath, gsPath: Uri, traceId: TraceId): IO[Unit] = io.file.readAll[IO](Paths.get(localObjectPath.asString), blockingEc, 4096).compile.to[Array].flatMap {
+    body =>
+      for {
+        bucketNameAndObjectName <- IO.fromEither(parseGsDirectory(gsPath.renderString).leftMap(InternalException))
+        _ <- googleStorageService.storeObject(bucketNameAndObjectName.bucketName, bucketNameAndObjectName.blobName, body, gcpObjectType, Some(traceId))
```

Comment:
Per the design doc, please pass the x-goog-if-generation-match header with the current known generation version. This could potentially be a follow-up change.

It would also be better to cache this metadata in memory so you don't need to refetch the GCS metadata on each delocalize call (this will slowdown the application).

Reply:
where does welder get generation version from?
We're using google's java lib (grpc), so it wouldn't be a header, but I think we do have option to pass it a generation variable

Reply:
It should be returned as part of the GCS object metadata. At the transport level, it's a header - but specifically for the Google Java Lib you will probably need `BlobTargetOption.generationMatch(XXX)`

Reply:
This is for storing object to gcs, so we don't have metadata for the object yet. Or am I missing something? Speaking of which, is /safeDelocalize supposed to adding metadata like lastEditedBy etc as well?

Reply:
If the object doesn't exist, we should pass generation of 0. This method will also be used for subsequent updates to the file from the Jupyter UI, in which case we want to send the previous known generation ID.

Good catch - I didn't call it out here, but yes it should also be adding the lastEdited/lastEditedBy metadata.

Reply:
do you mean before we delocalize, we'll always do a getMetadata call to get generation, increment that, and use that as the new generation number?

Reply:
Yes - except I would use the cached value here. There's no reason to refetch the metadata before delocalizing (unless we've never fetched the metadata before).

The generation number is just a value returned by GCS - we would need to refetch it at some point and should not attempt to increment it manually. More here: https://cloud.google.com/storage/docs/generations-preconditions

Reply:
(probably it would look like: 1. update 2. invalidate metadata cache for that file)